src/os2datascanner/utils/optimiser/optimiser.py

```diff
@@ -0,0 +1,314 @@
+    def clean_rule(self):
+        return load_json(self.OUT_PATH)
+
+    def setup(self):
+        """Copies the original rule into
+        another file where it is cleaned"""
+        obj = load_json(self.IN_PATH, extensive=True)
+        dump_json(obj, OUT_PATH)
+
+    def containify(self, container: 'CustomContainer'):
+        """Transforms all {"type": "and" | "or", "components": [...]} 
+        into the CustomContainer class, so they become easier to
+        handle in python, rather than dictionaries. Not k becomes -k.
+
+        *JSON to CustomContainer*"""
+
+        for k in container.components:
+            if isinstance(k, dict):
+                if k["type"] == "not":
+                    container.components = arr_switch(arr=container.components, deletion=k, insertion=-k["rule"])
+                else:
+                    new = CustomContainer(k["type"], k["components"], container)
+                    container.components = arr_switch(arr=container.components, deletion=k, insertion=new)
+                    self.containers.append(container.components[-1])
+                    yield container.components[-1]
+
+
+    def get_containers(self, main: 'CustomContainer'):
+        """Gets a list of all the containers.
+        Repeats itself with containify, but
+        otherwise the code wouldn't work."""
+        found = [main]
+        next_up: list[CustomContainer] = [main]
+        while next_up:
+            for cont in next_up:
+                for comp in cont.components:
+                    if isinstance(comp, dict):
+                        new = CustomContainer(comp["type"], comp["components"], cont)
+                        found.append(new)
+                        cont.components.remove(comp)
+                        cont.components.append(new)
+                        self.containers.append(new) # [new piece of code]
+                    elif isinstance(comp, CustomContainer):
+                        found.append(comp)
+                next_up.remove(cont)
+        return found
+
+
+    def check_empty(self, container:'CustomContainer'):
+        # Checks if container has no components i.e. AND(null) | OR(null)
+        if not container.components:
+            try:
+                container.parent.components.remove(container)
+                self.containers.remove(container)
+                self.found_redundancy = True
+            except ValueError:
+                pass
+            return True
+        return False
+
+
+    def check_useless(self, container:'CustomContainer'):
+        # Useless i.e. AND(1), OR(1), etc ...
+        if len(container.components) == 1:
+            try:
+                container.parent.components = arr_switch(container.parent.components, container, container.components[0])
+                self.containers.remove(container)
+                self.found_redundancy = True
+            except ValueError:
+                pass
+            return True
+        return False
```

Comment:
The `os2datascanner.engine2.rules` package already handles this case. If you try to make a `CompoundRule` with a single component, it automatically decays to that component:

```python
    @classmethod
    @abstractmethod
    def make(cls, *components):
        """Creates a new Rule that represents the combination of all of the
        given components. The result does not need to be an instance of @cls:
        it could be a completely different kind of Rule, or a simple boolean
        value, if this rule can already be reduced to one.

        Subclasses must override this method, but should call up to this
        implementation."""
        if len(components) == 0:
            raise ValueError("CompoundRule with zero components")
        elif len(components) == 1:
            return components[0]
        else:
            return cls(*components)
```

That is, we already implicitly do this optimisation when we execute the rule.

(That's not to say necessarily that you _shouldn't_ do it elsewhere too, but what if this was deliberate? What if you define a rule with an `AndRule` to leave yourself a hook to add other things later?)

Reply:
Your point here is that this test is useless, because the rule_builder does it before compiling it and sending to be checked at the scanner ? If so i can just delete my check ?